test/widgets/test_tasklist.py

```diff
@@ -0,0 +1,179 @@
+# Copyright (c) 2023 elParaguayo
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import pytest
+
+import libqtile.config
+from libqtile import bar, layout
+from libqtile.config import Screen
+from libqtile.confreader import Config
+from libqtile.widget.tasklist import TaskList, has_xdg
+
+
+@pytest.fixture
+def override_xdg(request):
+    return getattr(request, "param", False)
+
+
```

Comment:
Could we avoid having attributes at normal runtime that only exist for tests? We could monkey patch the widget here and uses that for tests to get the data that we want, perhaps something like:
```python
class TestTaskList(TaskList):
    def __init__(self, *args, **kwargs):
        TaskList.__init__(self, *args, **kwargs)
        self._text = ""

    def calc_box_widths(self):
        ret_val = TaskList.calc_box_widths(self)
        self._text = "|".join(self.get_taskname(w) for w in windows)
        return ret_val

    def info(self):
        info = base._Widget.info(self)
        info["text"] = self._text
        return info
```

Reply:
Absolutely. Nice idea.